spring/spring6/spring6-common/src/main/java/org/infinispan/spring/common/provider/SpringCache.java

```diff
@@ -259,4 +265,38 @@ static RuntimeException throwValueRetrievalException(Object key, Callable<?> loa
    public long getWriteTimeout() {
       return writeTimeout;
    }
+
+   @Override
+   public CompletableFuture<?> retrieve(Object key) {
+      return encodedToValueWrapper(nativeCache.getAsync(key));
+   }
+
+   @Override
+   public <T> CompletableFuture<T> retrieve(Object key, Supplier<CompletableFuture<T>> valueLoaderAsync) {
+
+      return nativeCache.getAsync(key).thenApply(v1 -> {
+         if (v1 == null) {
+            ReentrantLock lock = synchronousGetLocks.computeIfAbsent(key, k -> new ReentrantLock());
+            lock.lock();
+            try {
+               return nativeCache.getAsync(key).thenApply(v2 -> {
```

Comment:
Should this one also be a `.thenCompose`? I think it could end releasing the lock too early.